src/wallet/wallet2.cpp

```diff
@@ -4983,129 +4948,126 @@ void wallet2::restore(const std::string& wallet_, const epee::wipeable_string& p
+  multisig_account.m_aggregate_pubkey = rct::rct2pk(rct::identity());
+  multisig_account.m_base_privkey = blinded_skey;
+  multisig_account.m_base_ancillary_privkey = cryptonote::get_multisig_blinded_secret_key(get_account().get_keys().m_view_secret_key);
+  multisig_account.m_kex_rounds_complete = 0;
 
-      // Preparing data for the last round to calculate common public spend key. The data contains public multisig keys.
-      extra_multisig_info = pack_multisignature_keys(MULTISIG_EXTRA_INFO_MAGIC, secret_keys_to_public_keys(multisig_keys), rct::rct2sk(spend_skey));
-    }
-    else
-    {
-      // M / N case
-      MINFO("Preparing keys for next exchange round...");
+  cryptonote::multisig_kex_account_update(expanded_msgs, multisig_account);
 
-      // Preparing data for middle round - packing new public multisig keys to exchage with others.
-      extra_multisig_info = pack_multisignature_keys(MULTISIG_EXTRA_INFO_MAGIC, derivations, m_account.get_keys().m_spend_secret_key);
-      spend_skey = rct::sk2rct(m_account.get_keys().m_spend_secret_key);
+  CHECK_AND_ASSERT_THROW_MES(multisig_account.m_kex_rounds_complete == 1,
+    "'Rounds complete' has unexpected value while making multisig.");
 
-      // Need to store middle keys to be able to proceed in case of wallet shutdown.
-      m_multisig_derivations = derivations;
-    }
-  }
-  
+  // update wallet state
   if (!m_original_keys_available)
   {
     // Save the original i.e. non-multisig keys so the MMS can continue to use them to encrypt and decrypt messages
     // (making a wallet multisig overwrites those keys, see account_base::make_multisig)
-    m_original_address = m_account.get_keys().m_account_address;
-    m_original_view_secret_key = m_account.get_keys().m_view_secret_key;
+    m_original_address = get_account().get_keys().m_account_address;
+    m_original_view_secret_key = get_account().get_keys().m_view_secret_key;
     m_original_keys_available = true;
   }
 
   clear();
-  MINFO("Creating view key...");
-  crypto::secret_key view_skey = cryptonote::generate_multisig_view_secret_key(get_account().get_keys().m_view_secret_key, view_keys);
 
+  // account base
   MINFO("Creating multisig address...");
-  CHECK_AND_ASSERT_THROW_MES(m_account.make_multisig(view_skey, rct::rct2sk(spend_skey), rct::rct2pk(spend_pkey), multisig_keys),
-      "Failed to create multisig wallet due to bad keys");
-  memwipe(&spend_skey, sizeof(rct::key));
+  CHECK_AND_ASSERT_THROW_MES(m_account.make_multisig(multisig_account.m_aggregate_ancillary_privkey,
+    blinded_skey,
+    multisig_account.m_aggregate_pubkey,
+    multisig_account.m_privkeys),
+      "Failed to create multisig wallet account due to bad keys");
 
   init_type(hw::device::device_type::SOFTWARE);
   m_original_keys_available = true;
   m_multisig = true;
   m_multisig_threshold = threshold;
-  m_multisig_signers = multisig_signers;
-  ++m_multisig_rounds_passed;
+  m_multisig_signers = signers;
```

Comment:
I don't see how these two values can change from above, but it may be the diff messing with me.

Reply:
This is `make_multisig()`, so these values are uninitialized before this point.

Reply:
I thought this was setting the values in `multisig_account` again, but these are values in the `wallet2` class with the same name.